docs/internal/abi.md

```diff
@@ -0,0 +1,444 @@
+# Verona ABI
+
+## WARNING
+
+This is a preliminary document under heavy discussion.
+The ideas here are mostly right, the details, not necessarily.
+There will be a tutorial document that will be correct in all counts, but this isn't it.
+
+## Overview
+
+This document describes the representation of objects in Verona.
+Objects can be of various types, including numeric (`U32`, `F64`, `Bool`), pointers to interfaces or classes, type unions, etc.
+Depending on the context, objects can have a different representation (ex. inside unions).
+Objects can also be passed to functions or inserted into arrays, and this ABI defines how that works so that readers and writers, callers and callees all agree on the same representation.
+
+The following ABI proposal is under construction and will change over time.
+The public ABI (when interfacing with foreign languages or exported symbols) will be stable at some point and guaranteed by the compiler.
+But the compiler reserves the right to change the internal ABI at any time, where it can guarantee the changes are not visible from user code.
+
+## Memory Model
+
+Verona's memory model is based on isolated regions.
+References to sentinel objects (`iso`) point to the entry point of the region, while mutable references (`mut`) point to objects inside the region.
+Immutable references (`imm`) point to immutable objects outside of any region.
+The Verona runtime guarantees no more than one behaviour has access to the same region at any time, avoiding concurrent mutation problems.
+
+### Regions
+
+A region is a tree of related objects, not necessarily contiguous in memory.
+The reason for grouping objects in the same region is to they can be accessed together by the same behaviour.
+However, some regions (ex. sandboxes) do require memory to be contiguous (for range access protection).
+
+Two regions can have overlapping memory addresses.
+The Verona compiler guarantees that one behaviour cannot write to another region's memory by not having pointer arithmetic and checking provenance at compile time.
+However, dynamic arrays can still have dynamic out-of-bounds access, which should trigger a run time error.
+
+Objects have their sizes known statically and are allocated directly and assigned to a region.
+Dynamic arrays can grow their storage size, to add more elements at run time, but each element still has a static size.
+
+### Forests
+
+Regions can point to other regions (via their sentinel objects), recursively.
+Sentinel objects can only have a single owning reference, so when another region acquires the owning reference, it is moved from the previous region to the new one.
+This forms a forest, or a tree of trees of objects (not a DAG).
```

Comment:
My thoughts exactly, but the name Sylvan keeps repeating is "tree", so I used it.

Probably best if I just drop any attempts at nomenclature and just describe what things are.

Reply:
I think this is due to confusion over cowns and regions:

 - A region contains a *graph* of objects where every live object is dominated by the sentinel (and every non-live object is unreachable from the sentinel).
 - Regions are a arranged in a *tree* where an `iso` pointer in one region points to a sentinel in a child region.  A single region can have an unbounded (except by physical constraints of the machine) number of child regions but because there may only be one `iso` pointer from outside the region to the sentinel, regions may not form any graph structure other than a tree.
 - Each `cown` forms the root of a tree of regions.  The set of cowns defines a *forest* of regions.

Reply:
Excellent! I'll add that! Thanks!